### PR TITLE
[WIP] Move mb encoding specific functions to mbyte.

### PR DESCRIPTION
**Pull request by Yamakaky**
mbyte.c should only contain mb_\* functions.

I had to add the include in main.c because of the global function pointers.
Maybe I should put them in mbyte.h.

I added some functions in mbyte/\* to encapsulate.

There is a lot of work to do, I see a lot of `if(enc_utf8)` and co in the
code.

### CODE REVIEW

**Comment by bfredl** [comment]
Am not at computer so not sure if relevant, but keep #2905 in mind so you don't duplicate or go against it too much.

**Comment by Yamakaky** [comment]
Well, his PR is a change while mine is a refactor. I prefer his, though.

**Comment by ZyX-I** [comment]
I think that this should wait until #2159.

File: src/nvim/mbyte.c [hunk 1]
```diff
@@ -101,56 +101,14 @@
...
-struct interval {
-  long first;
-  long last;
-};
-
 #ifdef INCLUDE_GENERATED_DECLARATIONS
 # include "mbyte.c.generated.h"
+# include "mbyte/dbcs.h.generated.h"
+# include "mbyte/latin.h.generated.h"
+# include "mbyte/utf8.h.generated.h"
```
Comment (ZyX-I) [inline src/nvim/mbyte.c]: You should include `mbyte/dbcs.h` (and others) above with regular includes and include `.h.generated.h` files *there* and not here.

Comment (ZyX-I) [inline src/nvim/mbyte.c]: Given how `os/os.h` is written it is also acceptable to use `mbyte/mbyte.h` in place of three `.h` files.

File: src/nvim/main.c [hunk 2]
```diff
@@ -72,6 +72,10 @@
 #include "nvim/api/private/helpers.h"
 #include "nvim/api/private/handle.h"
 
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "mbyte/latin.h.generated.h"
+#endif
```
Comment (ZyX-I) [inline src/nvim/main.c]: This also should not be included here.

Comment (ZyX-I) [inline src/nvim/main.c]: No, this *must not* be included. `latin.h.generated.h` does *not* contain necessary `#include`s which defines things like `char_u`, they were always supposed to be written in corresponding `.h` file. By allowing such code you add *yet another* set of headers that cannot be used on their own. Ideal situation is when all headers can be included without the errors regardless of absence/presence of other headers’ includes, but this currently is not the case for many headers.

**Comment by justinmk** [comment]
@Yamakaky are you going to continue this? https://github.com/neovim/neovim/pull/2905 and  #2159 were merged.

**Comment by Yamakaky** [comment]
No, as utf8 by default seems a better option to me.

**Pull request closed without merging**